### PR TITLE
Refactor UserPrincipalManager to derive Key, resolve #472.

### LINKED ISSUE #472: Fixed algorithm when setup JWT token process
```java
final JWSAlgorithm expectedJWSAlg = JWSAlgorithm.RS256;
```

Should obtain this from JWT header.

### PR DESCRIPTION
**Pull request by DarrenForsythe**
## Summary
resolve #472.

Update the UserPrincipalManager to select the Algorithm based of
the JWT token. Inits the JWKSet in the AADAuthenticationFilter bean
construction to avoid costly I/O.

Add UserPrincipalManagerTest, jwt, and key now that the Keyset is a param
The JWT, and keys are generated from http://kjur.github.io/jsjws/tool_jwt.html.

## Issue Type

- New Feature

## Starter Names

  - active directory spring boot starter

## Additional Information

### CODE REVIEW

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/UserPrincipalManager.java [hunk 1]
```diff
@@ -7,37 +7,33 @@
...
-        this.serviceEndpoints = serviceEndpoints;
-        this.validator = getAadJwtTokenValidator();
-        this.resourceRetriever = resourceRetriever;
+    public UserPrincipalManager(JWKSource<SecurityContext> keySource) {
+        this.keySource
+                = keySource;
     }
 
     public UserPrincipal buildUserPrincipal(String idToken) throws ParseException, JOSEException, BadJOSEException {
+        final JWSHeader jwsHeader = JWSHeader.parse(new String(
```
Comment (Incarnation-p-lee) [inline azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/UserPrincipalManager.java]: As line 40, `jwsObject` will be parsed, you can parse the id_token first, and then reference the algo like `jwsObject.getHeader().getAlgoritm()`.

So the whole process may looks like below, `getAADB2CKeySource()` will create the keySource (As I mentioned below 
 to keep this uri as constructor parameter) and then you may avoiding parse the header twice.
```java
private final ConfigurableJWTProcessor<SecurityContext> processor = new DefaultJWTProcessor<>();
final JWSObject jwsObject = JWSObject.parse(idToken);
final JWSAlgorithm jwsAlgorithm = jwsObject.getHeader().getAlgorithm();

processor.setJWSKeySelector(new JWSVerificationKeySelector<>(jwsAlgorithm, getAADB2CKeySource()));

final JWTClaimsSet claimsSet = processor.process(idToken, null);

processor.getJWTClaimsSetVerifier().verify(claimsSet, null);
```

**Comment by Incarnation-p-lee** [comment]
This PR also resolve #472

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/UserPrincipalManager.java (hunk 1 above)
Comment (DarrenForsythe) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/UserPrincipalManager.java]: Updated to move the parsing up, remove parsing of header.

File: azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAuthenticationFilter.java [hunk 2]
```diff
@@ -37,22 +42,26 @@
...
                                    ResourceRetriever resourceRetriever) {
         this.aadAuthProps = aadAuthProps;
         this.serviceEndpointsProps = serviceEndpointsProps;
-        this.resourceRetriever = resourceRetriever;
-        this.principalManager = new UserPrincipalManager(
-                serviceEndpointsProps.getServiceEndpoints(aadAuthProps.getEnvironment()), resourceRetriever);
+        try {
+            final JWKSource<SecurityContext> keySource = new RemoteJWKSet<>(new URL(serviceEndpointsProps
+                    .getServiceEndpoints(aadAuthProps.getEnvironment()).getAadKeyDiscoveryUri()), resourceRetriever);
+            this.principalManager = new UserPrincipalManager(keySource);
```
[the remark this answers is not in the document]
Comment (DarrenForsythe) [reply azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/aad/AADAuthenticationFilter.java]: Re-added this constructor. Kept the new keysource constructor to allow for effective unit testing.

**Pull request merged**